### PR TITLE
Rosetta Constructions Endpoints

### PR DESCRIPTION
**Pull request by fosskers**
I will be adding more robust `NetworkId` checking in the next PR.

### CODE REVIEW

**Comment by fosskers** [comment]
Not quite done.

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 1]
```diff
@@ -77,13 +88,42 @@ someRosettaServer (FromSingChainwebVersion (SChainwebVersion :: Sing vT)) ms =
...
+        validated <- hoistEither . first (const RosettaInvalidTx) $ validateCommand cmd
+        mp <- (readMaybe (T.unpack n) >>= flip lookup ms) ?? RosettaInvalidChain n
+        let !vec = V.singleton validated
+        liftIO (mempoolInsertCheck mp vec) >>= hoistEither . first (const RosettaInvalidTx)
+        liftIO (mempoolInsert mp UncheckedInsert vec)
+        let rk = requestKeyToB16Text $ cmdToRequestKey validated
+        pure $ ConstructionSubmitResp (TransactionId rk) Nothing
+
+command :: T.Text -> Maybe (Command T.Text)
+command = decode' . TL.encodeUtf8 . TL.fromStrict
```
Comment (larskuhtz) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:120]:
```suggestion
command = decodeStrict' . T.encodeUtf8
```

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 2]
```diff
@@ -19,12 +21,19 @@ import Control.Error.Util
...
 
+import Data.Aeson
+import Data.Bifunctor
 import qualified Data.ByteString.Short as BSS
+import qualified Data.HashMap.Strict as HM
 import Data.Proxy (Proxy(..))
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.Encoding as TL
```
Comment (larskuhtz) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:32]:
```suggestion
```

See suggestion below

File: src/Chainweb/Pact/RestAPI/Server.hs [hunk 3]
```diff
@@ -1,4 +1,3 @@
-{-# LANGUAGE RecordWildCards #-}
```
Comment (larskuhtz) [inline src/Chainweb/Pact/RestAPI/Server.hs:1]: 👍

File: src/Chainweb/RestAPI.hs [hunk 4]
```diff
@@ -239,14 +241,15 @@ someChainwebServer v dbs mr (HeaderStream hs) (Rosetta r) =
...
-        <> bool mempty (someRosettaServer v mempools) r
+        <> bool mempty (someRosettaServer v concreteMs) r
   where
     payloads = _chainwebServerPayloadDbs dbs
     blocks = _chainwebServerBlockHeaderDbs dbs
     pacts = _chainwebServerPactDbs dbs
     cuts = _chainwebServerCutDb dbs
     peers = _chainwebServerPeerDbs dbs
     mempools = _chainwebServerMempools dbs
+    concreteMs = map (second (_chainResMempool . snd)) pacts
```
Comment (larskuhtz) [inline src/Chainweb/RestAPI.hs:252]: What about a more meaningful name and possibly a comment why `mempools` doesn't work here?

```suggestion
    concreteMempools = map (second (_chainResMempool . snd)) pacts
```

Comment (fosskers) [reply src/Chainweb/RestAPI.hs:252]: Lars and I are discussing this in person.

**Pull request merged**